### PR TITLE
Linux compile

### PR DESCRIPTION
**Pull request by thedavidharris**
Resolves https://github.com/tuist/xcodeproj/issues/296. Opening as a draft for some early feedback.

### Short description 📝
Draft PR of what changes are required to get the project compiling on Linux.

### Solution 📦
There were a few things that needed changing:
- Move back to Codable to generate a generic dictionary from the project XML, which required some type-erasure around the `Any` type so that it could be decoded.
- Removal of all ObjC references, some unused (like dictionary extensions) and NSObject usage.
- Removal of all Sourcery generated `isEqual` methods
- A copy of PropertyListEncoder into the project for Linux usage
- Macros to remove usage of MacOS-specific log functions
- Add test runners for Linux main

I can't quite get this to pass _all_ tests because I get the following in the Docker image I've been running:
```
Cloning into '/tmp/WWDC'...
fatal: unable to access 'https://github.com/insidegui/WWDC/': gnutls_handshake() failed: The TLS connection was non-properly terminated.
```

### Leftover
Leftover additional steps _at a minimum_. This definitely needs some work.
- [ ] Instead of commenting out CVarArg code, replace with a method functional on all platforms
- [ ] Verify integration tests
- [ ] Add Linux tests to CircleCI

### CODE REVIEW

File: Sources/xcodeproj/Objects/BuildPhase/PBXBuildPhase.swift [hunk 1]
```diff
@@ -114,7 +114,7 @@ public extension PBXBuildPhase {
     /// - Parameter file: file element to be added to the build phase.
     /// - Returns: proxy build file.
     /// - Throws: an error if the file cannot be added.
-    public func add(file: PBXFileElement) throws -> PBXBuildFile {
+    func add(file: PBXFileElement) throws -> PBXBuildFile {
```
Comment (pepibumur) [inline Sources/xcodeproj/Objects/BuildPhase/PBXBuildPhase.swift]: Out of curiosity, why was this method made internal? I'd keep it as public.

File: Sources/xcodeproj/Objects/Files/PBXFileElement.swift [hunk 2]
```diff
@@ -140,7 +140,7 @@ public extension PBXFileElement {
     /// - Parameter sourceRoot: project source root.
     /// - Returns: file element absolute path.
     /// - Throws: an error if the absolute path cannot be obtained.
-    public func fullPath(sourceRoot: Path) throws -> Path? {
+    func fullPath(sourceRoot: Path) throws -> Path? {
```
Comment (pepibumur) [inline Sources/xcodeproj/Objects/Files/PBXFileElement.swift]: Same here and with all the methods that have been made internal

File: Sources/xcodeproj/Objects/Project/PBXObjectReference.swift [hunk 3]
```diff
@@ -70,9 +72,8 @@ class PBXObjectReference: NSObject, Comparable, NSCopying {
         }
     }
 
-    /// Hash value.
-    override var hash: Int {
-        return value.hashValue
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(value)
```
Comment (pepibumur) [inline Sources/xcodeproj/Objects/Project/PBXObjectReference.swift]: 👏

File: Sources/xcodeproj/Scheme/XCScheme+ArchiveAction.swift [hunk 4]
```diff
@@ -47,13 +47,15 @@ extension XCScheme {
...
-        override func isEqual(to: Any?) -> Bool {
-            guard let rhs = to as? ArchiveAction else { return false }
-            return super.isEqual(to: to) &&
-                buildConfiguration == rhs.buildConfiguration &&
-                revealArchiveInOrganizer == rhs.revealArchiveInOrganizer &&
-                customArchiveName == rhs.customArchiveName
+        
+
+        static func == (lhs: ArchiveAction, rhs: ArchiveAction) -> Bool {
+            return
```
Comment (pepibumur) [inline Sources/xcodeproj/Scheme/XCScheme+ArchiveAction.swift]: Notice that we got rid of the `super.isEqual(to: to)` that delegates the comparison of some attributes to its parents.

File: Sources/xcodeproj/Scheme/XCScheme+LaunchAction.swift [hunk 5]
```diff
@@ -272,40 +272,5 @@ extension XCScheme {
             }
             return element
         }
-
-        // MARK: - Equatable
-
-        override func isEqual(to: Any?) -> Bool {
```
Comment (pepibumur) [inline Sources/xcodeproj/Scheme/XCScheme+LaunchAction.swift]: Am I wrong in assuming that the conformance to `Equatable` is not generated automatically by the compiler?

File: Sources/xcodeproj/Objects/BuildPhase/PBXBuildPhase.swift (hunk 1 above)
Comment (thedavidharris) [reply Sources/xcodeproj/Objects/BuildPhase/PBXBuildPhase.swift]: They are still public, they're inside a `public` extension so once you switch to the Swift 5 toolchain, it'll give a warning saying that the modifier is redundant. I'm fine with reverting these.

File: Sources/xcodeproj/Scheme/XCScheme+ArchiveAction.swift (hunk 4 above)
Comment (thedavidharris) [reply Sources/xcodeproj/Scheme/XCScheme+ArchiveAction.swift]: Yeah, polymorphic equality gets tricky when you move away from NSObject. I mostly tried to manually check the elements on the superclass as well which is not ideal; I think casting and using the superclass operator may work though.

File: Sources/xcodeproj/Scheme/XCScheme+LaunchAction.swift (hunk 5 above)
Comment (thedavidharris) [reply Sources/xcodeproj/Scheme/XCScheme+LaunchAction.swift]: Not for classes and tuples: https://github.com/apple/swift-evolution/blob/master/proposals/0185-synthesize-equatable-hashable.md#synthesis-for-class-types-and-tuples.

A number of these classes could probably be structs though.

**Comment by dharr419-ford** [comment]
```
 let serialized = try PropertyListSerialization.propertyList(
            from: plistXML,
            options: .mutableContainersAndLeaves,
            format: &propertyListFormat
        )
```

doesn't compile on Linux with an error of `/app/Sources/xcodeproj/Project/XcodeProj.swift:88:21: error: extra argument 'format' in call
            format: &propertyListFormat` but...

```
 let serialized = try PropertyListSerialization.propertyList(
            from: plistXML,
            options: [],
            format: &propertyListFormat
        )
```

does. 🤔 

Couldn't find any particular reason why.

**Comment by dharr419-ford** [comment]
Added a Linux job to CI. Docker _really_ doesn't seem to like the git clone though, it continues to throw a `fatal: unable to access 'https://github.com/insidegui/WWDC/': gnutls_handshake() failed: The TLS connection was non-properly terminated.` Not sure how to fix that one. 

Still need to verify some of the other steps and clean this up, and it required removing the `.mutableContainersAndLeaves,` from the Plist serialization call which didn't seem to affect much. Also not sure on how nice it is to include an entire file from the stdlib inside the project itself, so will think more on that.

--David

**Comment by thedavidharris** [comment]
Alright, everything seems to be building fine and I've added a CI task for Linux builds as well. Still leaving as a draft for the remaining outstanding issues:

- Need to fix up the `Equatable`-ness and figure out what to do with the Sourcery template.
- Linux tests continue to fail because of Docker issues. CI is giving something cryptic, but locally when running in the same Docker image I get:
```
Test Suite 'OSSProjectsTests' started at 2019-02-24 00:14:40.900
Test Case 'OSSProjectsTests.test_projects' started at 2019-02-24 00:14:40.900
Cloning into '/tmp/WWDC'...
fatal: unable to access 'https://github.com/insidegui/WWDC/': gnutls_handshake() failed: The TLS connection was non-properly terminated.
> Cloning https://github.com/insidegui/WWDC to run the integration test
<EXPR>:0: error: OSSProjectsTests.test_projects : threw error "Command '/usr/bin/git clone --depth=1 https://github.com/insidegui/WWDC /tmp/WWDC' returned with error code 128."
Test Case 'OSSProjectsTests.test_projects' failed (0.208 seconds)
Test Suite 'OSSProjectsTests' failed at 2019-02-24 00:14:41.108
	 Executed 1 test, with 1 failure (1 unexpected) in 0.208 (0.208) seconds
```

@mxcl I know this isn't a Linux Swift particular thing, but would you happen to know?

Also, would be good to make sure that the test Linux main is generated automatically with any changes to the tests.

**Comment by keith** [comment]
That error kinda looks like a networking flake. Maybe restart them? I ran the tests from this branch on a Linux machine and they at least didn't have a problem cloning it:

```
Test Suite 'OSSProjectsTests' started at 2019-02-24 00:20:07.157
Test Case 'OSSProjectsTests.test_projects' started at 2019-02-24 00:20:07.157
> Cloning https://github.com/insidegui/WWDC to run the integration test
Cloning into '/tmp/WWDC'...
remote: Enumerating objects: 655, done.
remote: Counting objects: 100% (655/655), done.
remote: Compressing objects: 100% (600/600), done.
remote: Total 655 (delta 90), reused 372 (delta 40), pack-reused 0
Receiving objects: 100% (655/655), 32.84 MiB | 32.09 MiB/s, done.
Resolving deltas: 100% (90/90), done.
<EXPR>:0: error: OSSProjectsTests.test_projects : threw error "Could not execute file at path 'cd'."
```

**Comment by thedavidharris** [comment]
@keith yeah that's the same one that gets thrown on CircleCI.

Also for what it's worth, there seems to be a bit of a gap in the unit tests around the equatablility of objects; tests pass here, but this does break XcodeGen.

**Comment by pepibumur** [comment]
>Alright, everything seems to be building fine and I've added a CI task for Linux builds as well. Still leaving as a draft for the remaining outstanding issues:

This is neat. You have done a fabulous job with this PR 💪

Have you tried changing the test that is failing to output the standard error from the commands that are being run? This doesn't say much about what's going on in the Linux environment:

```
<EXPR>:0: error: OSSProjectsTests.test_projects : threw error "Command '/usr/bin/git clone --depth=1 https://github.com/insidegui/WWDC /tmp/WWDC' returned with error code 128."
```

> Also for what it's worth, there seems to be a bit of a gap in the unit tests around the equatablility of objects; tests pass here, but this does break XcodeGen.

After we introduced Sourcery to generate the Equatable conformances we didn't keep those tests up to date consistently. What exactly breaks?

**Comment by thedavidharris** [comment]
I'll post the test failures from XcodeGen and Tuist when I get a chance. Have to investigate that failure too, can't really even be sure everythign is working without that integration test.

For what it's worth, https://github.com/apple/swift-corelibs-foundation/commit/d2d72f88d93f7645b94c21af88a7c9f69c979e4f#diff-4ad636dfe600e16033f732565a0592fc dropped in the daily dev snapshots, so should undo the need from copying the whole file from the Darwin Foundation overlay and modifying it a bit. Will test that out as well.

**Comment by pepibumur** [comment]
> I'll post the test failures from XcodeGen and Tuist when I get a chance. Have to investigate that failure too, can't really even be sure everythign is working without that integration test.

I wouldn't be surprised if it has to do with the test being dependent on the macOS environment.

> For what it's worth, [apple/swift-corelibs-foundation@d2d72f8#diff-4ad636dfe600e16033f732565a0592fc](https://github.com/apple/swift-corelibs-foundation/commit/d2d72f88d93f7645b94c21af88a7c9f69c979e4f#diff-4ad636dfe600e16033f732565a0592fc) dropped in the daily dev snapshots, so should undo the need from copying the whole file from the Darwin Foundation overlay and modifying it a bit. Will test that out as well.

That's awesome. We can open an issue after this gets merge to remove it once that change makes it to an official release.

**Comment by pepibumur** [comment]
Hey @thedavidharris, anything that I can help you with on this PR? Happy to give you a hand to get this through.

**Comment by thedavidharris** [comment]
I've got some exams to study for coming up so won't be able to look at this for a bit, but think this captures the general state. All unit tests compile and run, but there are issues around removing all of the Sourcery generated `@objc override isEqual` functions since overriding that on NSObject-derived classes doesn't work. I'm not quite sure if the way I re-implemented the polymorphic equality was right, but I do know that plugging this into `XcodeGen` and running the unit tests with it causes a crash that obviously doesn't happen when using the current version. Haven't been able to look into it much.

**Comment by pepibumur** [comment]
>  I've got some exams to study for coming up so won't be able to look at this for a bit, but think this captures the general state. All unit tests compile and run, but there are issues around removing all of the Sourcery generated `@objc override isEqual` functions since overriding that on NSObject-derived classes doesn't work. I'm not quite sure if the way I re-implemented the polymorphic equality was right, but I do know that plugging this into `XcodeGen` and running the unit tests with it causes a crash that obviously doesn't happen when using the current version. Haven't been able to look into it much

No worries, no rush to get this one done. I'll investigate it and see if I can find anything. All the best with those exams, you'll nail them!

**Comment by pepibumur** [comment]
Hey @thedavidharris, apologies for not following up on this PR after my last comment. How did your exams go?
I was wondering if you are still interested in finishing up this work. We are planning a major release and I think this would be a great addition to the project.

**Comment by thedavidharris** [comment]
@pepibumur not bad! 

We have a bit less of a need for this as the main reason we wanted this for Linux is we have an internal product that does all the provisioning/bootstrapping of new mobile projects and wanted to see if we could leverage this on our existing Linux machines, but we ended up deploying it on Macs. 

It would still be pretty useful but might be better to turn this into a few smaller PRs with the less risky changes to work towards it. Most of the changes other than the `@objc override isEqual` should be doable, those are what's really holding this up.

**Comment by pepibumur** [comment]
> @pepibumur not bad!

That's great to hear 💪

> We have a bit less of a need for this as the main reason we wanted this for Linux is we have an internal product that does all the provisioning/bootstrapping of new mobile projects and wanted to see if we could leverage this on our existing Linux machines, but we ended up deploying it on Macs.

That explains your interest in how we configured the fleet at Shopify 😜. Do you have all those Macs already working?

> It would still be pretty useful but might be better to turn this into a few smaller PRs with the less risky changes to work towards it. Most of the changes other than the @objc override isEqual should be doable, those are what's really holding this up.

Agree. Would you like to lead this effort? If you don't feel like doing it I can take it over and do the split-up work.

**Comment by thedavidharris** [comment]
> That explains your interest in how we configured the fleet at Shopify 😜. Do you have all those Macs already working?

Not yet, working on it!

> Agree. Would you like to lead this effort? If you don't feel like doing it I can take it over and do the split-up work.

I'll refocus this PR to just include the less risky changes for now; it's still just a draft so should be alright I think. Then can just poke at the equality things one by one to figure out which one is behaving a little odd (unit tests pass here, but if you pull this fork into xcodegen you get some exceptions)

**Comment by pepibumur** [comment]
I'm closing this in favor of https://github.com/tuist/XcodeProj/pull/504

**Pull request closed without merging**